libvast/src/default_table_slice_builder.cpp

```diff
@@ -53,9 +57,10 @@ table_slice_ptr default_table_slice_builder::finish() {
   // TODO: this feels messy, but allows for non-virtual parent accessors.
   slice_->rows_ = slice_->xs_.size();
   slice_->columns_ = layout_.fields.size();
-  auto result = slice_;
-  slice_ = {};
-  return result;
+  using std::swap;
+  default_table_slice_ptr result;
+  swap(slice_, result);
+  return table_slice_handle{std::move(result)};
```

Comment:
Out of curiosity, why do you prefer the `swap` version over the explicit resetting?

Reply:
To me it reads much cleaner this way. Our intend is to take whatever is currently in `slice_` and return it. The construct-swap-return cycle has the exact semantic we are looking for and adds no extra noise or side effects.

The previous version did a copy of `slice_` only to throw away one copy. This is not elegant in my opinion and it also means unnecessary reference counting overhead that the compiler probably won't be able to optimize away (writes to the reference count are visible to other threads and the standard requires the compiler to produce all visible side effects that the abstract machine would). Rewriting the previous code to:

```
auto result = std::move(slice_);
slice_ = {};
return result;
```

Would solve the efficiency issue, but I still prefer the cleaner construct-swap over assign-assign.

Reply:
Great explanation, I'm on the same page now. Thanks for elaborating!

Reply:
Great. :)

On a side note, only for the sake of completeness: `auto result = std::move(slice_); return result;` would *technically* be enough if you take a look at how the move-constructor of `intrusive_ptr` is implemented. However, according to the standard this leaves `slice_` in a moved-from state and the only thing you can to to "revive" it safely is assigning into it. So even if that's the "most efficient" implementation, it would rely on implementation-defined behavior (in other words: dirty coding style).